aws/ecs_service.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+// filterOutRecentServices - Given a list of services and an excludeAfter
+// timestamp, filter out any services that were created after `excludeAfter`.
+func filterOutRecentServices(svc *ecs.ECS, clusterArn *string, ecsServiceArns []string, excludeAfter time.Time) ([]*string, error) {
+	// Fetch descriptions in batches of 10, which is the max that AWS
+	// accepts for describe service.
+	var filteredEcsServiceArns []*string
+	batches := split(ecsServiceArns, 10)
+	for _, batch := range batches {
+		params := &ecs.DescribeServicesInput{
+			Cluster:  clusterArn,
+			Services: awsgo.StringSlice(batch),
+		}
+		describeResult, err := svc.DescribeServices(params)
+		if err != nil {
+			return nil, errors.WithStackTrace(err)
+		}
+		for _, service := range describeResult.Services {
+			if excludeAfter.After(*service.CreatedAt) {
+				filteredEcsServiceArns = append(filteredEcsServiceArns, service.ServiceArn)
+			}
+		}
+	}
+	return filteredEcsServiceArns, nil
+}
+
+// getAllEcsServices - Returns a formatted string of ECS Service ARNs, which
+// uniquely identifies the service, in addition to a mapping of services to
+// clusters. For ECS, need to track ECS clusters of services as all service
+// level API endpoints require providing the corresponding cluster.
+// Note that this looks up services by ECS cluster ARNs.
+func getAllEcsServices(awsSession *session.Session, ecsClusterArns []*string, excludeAfter time.Time) ([]*string, map[string]string, error) {
+	ecsServiceClusterMap := map[string]string{}
+	svc := ecs.New(awsSession)
+
+	// For each cluster, fetch all services, filtering out recently created
+	// ones.
+	var ecsServiceArns []*string
+	for _, clusterArn := range ecsClusterArns {
+		result, err := svc.ListServices(&ecs.ListServicesInput{Cluster: clusterArn})
+		if err != nil {
+			return nil, nil, errors.WithStackTrace(err)
+		}
+		filteredServiceArns, err := filterOutRecentServices(svc, clusterArn, awsgo.StringValueSlice(result.ServiceArns), excludeAfter)
+		if err != nil {
+			return nil, nil, errors.WithStackTrace(err)
+		}
+		// Update mapping to be used later in nuking
+		for _, serviceArn := range filteredServiceArns {
+			ecsServiceClusterMap[*serviceArn] = *clusterArn
+		}
+		ecsServiceArns = append(ecsServiceArns, filteredServiceArns...)
+	}
+
+	return ecsServiceArns, ecsServiceClusterMap, nil
+}
+
+// Deletes all provided ECS Services. At a high level this involves two steps:
+// 1.) Drain all tasks from the service so that nothing is
+//     running.
+// 2.) Delete service object once no tasks are running.
+// Note that this will swallow failed deletes and continue along, logging the
+// service ARN so that we can find it later.
+func nukeAllEcsServices(awsSession *session.Session, ecsServiceClusterMap map[string]string, ecsServiceArns []*string) error {
```

Comment:
This is a really long function with big comment blocks in the middle. Consider breaking the contents into several smaller functions: e.g., `startDrainingEcsServices`, `waitForServicesToFinishDraining`, etc.

Reply:
Done